video/gazebocamerainput.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+// BoB robotics includes
+#include "common/assert.h"
+#include "input.h"
+
+// Gazebo includes
+#include <gazebo/gazebo_config.h>
+#include <gazebo/transport/transport.hh>
+#include <gazebo/msgs/msgs.hh>
+#include <gazebo/common/common.hh>
+
+// Gazebo's API has changed between major releases. These changes are
+// accounted for with #if..#endif blocks in this file.
+#if GAZEBO_MAJOR_VERSION < 6
+#include <gazebo/gazebo.hh>
+#else
+#include <gazebo/gazebo_client.hh>
+#endif
+
+// OpenCV
+#include <opencv2/opencv.hpp>
+
+// Standard C++ includes
+#include <stdexcept>
+
+namespace BoBRobotics {
+namespace Video {
+
+constexpr const char *GazeboCameraDeviceName = "Gazebo Camera";
+
+//----------------------------------------------------------------------------
+// BoBRobotics::Video::GazeboCameraInput
+//----------------------------------------------------------------------------
+//! A thin wrapper for reading from any video source supported by OpenCV
+class GazeboCameraInput : public Input
+{
+public:
+    //! Create a new video stream
+    GazeboCameraInput()
+      : GazeboCameraInput("/gazebo/default/my_simple_cart/camera/link/camera/image")
+    {}
+
+    /*!
+     * \brief Create a video stream for a specific device
+     *
+     * @param device Integer or string representation of device (passed to
+     *        cv::VideoCapture's constructor)
+     * @param cameraName The short name to use for this camera (see getCameraName())
+     */
+    
+    GazeboCameraInput(const std::string &topic, const std::string &cameraName = GazeboCameraDeviceName)
+      : m_CameraTopic(topic), m_CameraName(cameraName)
+    {
+        // Load gazebo as a client
+        #if GAZEBO_MAJOR_VERSION < 6
+        gazebo::setupClient(0, 0);
+        #else
+        gazebo::client::setup(0, 0);
+        #endif
+        //Setup image subscription node
+        m_ImageNode= gazebo::transport::NodePtr(new gazebo::transport::Node());
+        m_ImageNode->Init();
```

Comment:
A single node can be used for publish and subscribe but we would need the node object to be accessible from both the gazebocamerainput.h and simulated_tank_gazebo.cc files. How and where should it be defined?

Reply:
We could just let the user define it and then pass it in to the two classes. I'm guessing the ``gazebo::client::setup(0,0);`` business only needs to be done once as well, right?

You could wrap all this in a little ``getGazeboNode()`` function, but this might be overkill given that it's only a few lines of code.

Reply:
Yes, there can be a common header with this `getGazeboNode` defined, which may be useful in the future as well.
There's one more problem. The current implementation requires one to launch the simulated_tank_gazebo and display_test separately, in which case, there would anyway be 2 nodes at runtime, one for each executable. I'm guessing we need to write a driver application which would start both in separate threads and pass a single gazebo node to both. Is this correct?

Reply:
Yeah. Maybe a combined example with a driveable robot + display would be best. You don't necessarily need two threads though. Take a look at some of the other examples, e.g.: https://github.com/BrainsOnBoard/bob_robotics/blob/master/examples/robot_net/computer.cc (this one's for controlling a robot over the network)

You can usually do most things on the main thread, polling the appropriate functions in a loop.

Reply:
Was able to get the display running parallelly by calling display.runInBackground. Now we can run the camera and motor drive using a single gazebo node. Incorporated this into examples/simulated_tank_gazebo in the latest commit. Execute using `./run.sh -d`

Reply:
:+1: